diff -urN hadoop-3.3.4-src.orig/hadoop-common-project/hadoop-common/src/main/native/src/org/apache/hadoop/crypto/OpensslCipher.c hadoop-3.3.4-src/hadoop-common-project/hadoop-common/src/main/native/src/org/apache/hadoop/crypto/OpensslCipher.c
--- hadoop-3.3.4-src.orig/hadoop-common-project/hadoop-common/src/main/native/src/org/apache/hadoop/crypto/OpensslCipher.c	2022-05-09 18:24:14.000000000 +0200
+++ hadoop-3.3.4-src/hadoop-common-project/hadoop-common/src/main/native/src/org/apache/hadoop/crypto/OpensslCipher.c	2022-12-22 10:44:54.227098830 +0100
@@ -35,8 +35,14 @@
 #endif
 static int (*dlsym_EVP_CIPHER_CTX_set_padding)(EVP_CIPHER_CTX *, int);
 static int (*dlsym_EVP_CIPHER_CTX_test_flags)(const EVP_CIPHER_CTX *, int);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+static int (*dlsym_EVP_CIPHER_CTX_get_block_size)(const EVP_CIPHER_CTX *);
+#else
 static int (*dlsym_EVP_CIPHER_CTX_block_size)(const EVP_CIPHER_CTX *);
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+static int (*dlsym_EVP_CIPHER_CTX_is_encrypting)(const EVP_CIPHER_CTX *);
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
 static int (*dlsym_EVP_CIPHER_CTX_encrypting)(const EVP_CIPHER_CTX *);
 #endif
 static int (*dlsym_EVP_CipherInit_ex)(EVP_CIPHER_CTX *, const EVP_CIPHER *,  \
@@ -56,8 +62,14 @@
 typedef void (__cdecl *__dlsym_EVP_CIPHER_CTX_init)(EVP_CIPHER_CTX *);
 typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_set_padding)(EVP_CIPHER_CTX *, int);
 typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_test_flags)(const EVP_CIPHER_CTX *, int);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_get_block_size)(const EVP_CIPHER_CTX *);
+#else
 typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_block_size)(const EVP_CIPHER_CTX *);
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_is_encrypting)(const EVP_CIPHER_CTX *);
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
 typedef int (__cdecl *__dlsym_EVP_CIPHER_CTX_encrypting)(const EVP_CIPHER_CTX *);
 #endif
 typedef int (__cdecl *__dlsym_EVP_CipherInit_ex)(EVP_CIPHER_CTX *,  \
@@ -75,8 +87,14 @@
 static __dlsym_EVP_CIPHER_CTX_init dlsym_EVP_CIPHER_CTX_init;
 static __dlsym_EVP_CIPHER_CTX_set_padding dlsym_EVP_CIPHER_CTX_set_padding;
 static __dlsym_EVP_CIPHER_CTX_test_flags dlsym_EVP_CIPHER_CTX_test_flags;
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+static __dlsym_EVP_CIPHER_CTX_get_block_size dlsym_EVP_CIPHER_CTX_get_block_size;
+#else
 static __dlsym_EVP_CIPHER_CTX_block_size dlsym_EVP_CIPHER_CTX_block_size;
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+static __dlsym_EVP_CIPHER_CTX_is_encrypting dlsym_EVP_CIPHER_CTX_is_encrypting;
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
 static __dlsym_EVP_CIPHER_CTX_encrypting dlsym_EVP_CIPHER_CTX_encrypting;
 #endif
 static __dlsym_EVP_CipherInit_ex dlsym_EVP_CipherInit_ex;
@@ -141,9 +159,17 @@
                       "EVP_CIPHER_CTX_set_padding");
   LOAD_DYNAMIC_SYMBOL(dlsym_EVP_CIPHER_CTX_test_flags, env, openssl,  \
                       "EVP_CIPHER_CTX_test_flags");
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+  LOAD_DYNAMIC_SYMBOL(dlsym_EVP_CIPHER_CTX_get_block_size, env, openssl,  \
+                      "EVP_CIPHER_CTX_get_block_size");
+#else
   LOAD_DYNAMIC_SYMBOL(dlsym_EVP_CIPHER_CTX_block_size, env, openssl,  \
                       "EVP_CIPHER_CTX_block_size");
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+  LOAD_DYNAMIC_SYMBOL(dlsym_EVP_CIPHER_CTX_is_encrypting, env, openssl,  \
+                      "EVP_CIPHER_CTX_is_encrypting");
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
   LOAD_DYNAMIC_SYMBOL(dlsym_EVP_CIPHER_CTX_encrypting, env, openssl,  \
                       "EVP_CIPHER_CTX_encrypting");
 #endif
@@ -171,10 +197,20 @@
   LOAD_DYNAMIC_SYMBOL(__dlsym_EVP_CIPHER_CTX_test_flags,  \
                       dlsym_EVP_CIPHER_CTX_test_flags, env,  \
                       openssl, "EVP_CIPHER_CTX_test_flags");
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+  LOAD_DYNAMIC_SYMBOL(__dlsym_EVP_CIPHER_CTX_get_block_size,  \
+                      dlsym_EVP_CIPHER_CTX_get_block_size, env,  \
+                      openssl, "EVP_CIPHER_CTX_get_block_size");
+#else
   LOAD_DYNAMIC_SYMBOL(__dlsym_EVP_CIPHER_CTX_block_size,  \
                       dlsym_EVP_CIPHER_CTX_block_size, env,  \
                       openssl, "EVP_CIPHER_CTX_block_size");
-#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+#endif
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+  LOAD_DYNAMIC_SYMBOL(__dlsym_EVP_CIPHER_CTX_is_encrypting,  \
+                      dlsym_EVP_CIPHER_CTX_is_encrypting, env,  \
+                      openssl, "EVP_CIPHER_CTX_is_encrypting");
+#elif OPENSSL_VERSION_NUMBER >= 0x10100000L
   LOAD_DYNAMIC_SYMBOL(__dlsym_EVP_CIPHER_CTX_encrypting,  \
                       dlsym_EVP_CIPHER_CTX_encrypting, env,  \
                       openssl, "EVP_CIPHER_CTX_encrypting");
@@ -307,11 +343,17 @@
     }
     return 0;
   } else {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
     int b = dlsym_EVP_CIPHER_CTX_block_size(context);
+#else
+    int b = dlsym_EVP_CIPHER_CTX_get_block_size(context);
+#endif
 #if OPENSSL_VERSION_NUMBER < 0x10100000L
     if (context->encrypt) {
-#else
+#elif OPENSSL_VERSION_NUMBER < 0x30000000L
     if (dlsym_EVP_CIPHER_CTX_encrypting(context)) {
+#else
+    if (dlsym_EVP_CIPHER_CTX_is_encrypting(context)) {
 #endif
       if (max_output_len >= input_len + b - 1) {
         return 1;
@@ -366,7 +408,11 @@
   if (  dlsym_EVP_CIPHER_CTX_test_flags(context, EVP_CIPH_NO_PADDING) ) {
     return 1;
   } else {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
     int b = dlsym_EVP_CIPHER_CTX_block_size(context);
+#else
+    int b = dlsym_EVP_CIPHER_CTX_get_block_size(context);
+#endif
     if (max_output_len >= b) {
       return 1;
     }
